server/zanata-war/src/main/java/org/zanata/service/impl/CopyVersionServiceImpl.java

```diff
@@ -284,37 +286,73 @@ public int getTotalDocCount(@Nonnull String projectSlug,
      */
     private void copyTextFlowAndTarget(Long documentId, Long newDocumentId) {
         copyTfAndTftStopWatch.start();
-        int tfStart = 0, tftStart = 0, totalTftCount = 0;
+        int tfStart = 0;
+//        AtomicInteger totalTftCount = new AtomicInteger(0);
         int tfCount = textFlowDAO.countActiveTextFlowsInDocument(documentId);
 
         while (tfStart < tfCount) {
-            Map<Long, Long> tfMap =
-                    copyTextFlowBatch(documentId, newDocumentId, tfStart,
+            List<HTextFlow> textFlows = textFlowDAO
+                    .getTextFlowsByDocumentId(documentId, tfStart,
                             TF_BATCH_SIZE);
-            tfStart += TF_BATCH_SIZE;
-            textFlowDAO.clear();
-            documentDAO.clear();
-
-            for (Map.Entry<Long, Long> entry : tfMap.entrySet()) {
-                tftStart = 0;
-                int tftCount =
-                        textFlowTargetDAO.countTextFlowTargetsInTextFlow(
-                                entry.getKey());
-
-                while (tftStart < tftCount) {
-                    totalTftCount +=
-                            copyTextFlowTargetBatch(entry.getKey(),
-                                    entry.getValue(), tftStart, TFT_BATCH_SIZE);
-                    tftStart += TFT_BATCH_SIZE;
-                    textFlowDAO.clear();
-                    textFlowTargetDAO.clear();
-                }
+            try {
+                Map<Long, Long> copyTextFlowsResult = transactionUtil.call(() -> {
+                    Map<Long, HTextFlow> tfMap = Maps.newHashMap();
+
+                    HDocument newDocument = documentDAO.getById(newDocumentId);
+                    for (HTextFlow textFlow : textFlows) {
+                        HTextFlow newTextFlow =
+                                copyTextFlow(newDocument, textFlow);
+//                        newTextFlow.copyPos(textFlow);
+                        newDocument.getTextFlows().add(newTextFlow);
+                        textFlowDAO.makePersistent(newTextFlow);
+
+                        tfMap.put(textFlow.getId(), newTextFlow);
+                    }
+                    documentDAO.makePersistent(newDocument);
+                    documentDAO.flush();
+
+                    return Maps.transformEntries(tfMap,
+                            (key, value) -> value.getId());
+
+                });
+
+
+                tfStart += TF_BATCH_SIZE;
+                textFlowDAO.clear();
+                documentDAO.clear();
+
+                textFlows.forEach(tf -> {
+                    try {
+                        transactionUtil.run(() -> {
+                            HTextFlow newTextFlow = textFlowDAO
+                                    .findById(copyTextFlowsResult.get(tf.getId()));
+
```

Comment:
It's not better... It's pretty much an experiment to see whether this will make things better. I guess you missed all the discussions we had for this. There was a performance bug for copy version but can't reproduce. This came out of that investigation and trial. Sean suggested I should put this work in a branch so that it's not wasted. He or anyone else can have a look at it and see if anything else can be improved.